### PR TITLE
DO NOT MERGE fixes to ORSet mergeRemoveDelta and ORMap deltaMerge

### PR DESCRIPTION
**Pull request by gosubpl**
* fix incorrect merge of ORSet deltas (#22631)
* fix incorrect merge of ORMap deltas (#22648)

### CODE REVIEW

**Comment by gosubpl** [comment]
Refs #22631 
Refs #22648

File: akka-distributed-data/src/test/scala/akka/cluster/ddata/ORSetSpec.scala [hunk 1]
```diff
@@ -419,6 +419,40 @@ class ORSetSpec extends WordSpec with Matchers {
       t4.elements should ===(Set("b", "a"))
     }
 
+    "not have anomalies for ORSet in complex but realistic scenario" in {
```
Comment (gosubpl) [inline akka-distributed-data/src/test/scala/akka/cluster/ddata/ORSetSpec.scala]: Test from #22631

File: akka-distributed-data/src/test/scala/akka/cluster/ddata/ORMultiMapSpec.scala [hunk 2]
```diff
@@ -86,6 +86,31 @@ class ORMultiMapSpec extends WordSpec with Matchers {
     }
   }
 
+  "be able to have its entries correctly merged with another ORMultiMap with overlapping entries 2" in {
```
Comment (gosubpl) [inline akka-distributed-data/src/test/scala/akka/cluster/ddata/ORMultiMapSpec.scala]: Test from #22648

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala [hunk 3]
```diff
@@ -383,12 +383,17 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
       case putOp: PutDeltaOp[A, B] ⇒
         val key = putOp.value._1
         thatValueDeltas += (key → (putOp.value :: Nil)) // put is destructive!
-      case _: RemoveDeltaOp[A, B] ⇒
-      // remove delta is only for the side effect of key being removed
-      // please note that if it is not preceded by update clearing the value
-      // anomalies will result
+      case removeOp: RemoveDeltaOp[A, B] ⇒
```
Comment (gosubpl) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala]: Fix for #22648 but would be incorrect without corresponding fix in `ORSet` below

**Comment by gosubpl** [comment]
Will rebase tomorrow, plus there is a small error in one of the unit tests - the expected outcome is incorrect, but the whole test is ok.

**Comment by patriknw** [comment]
sounds great, I'll review later this evening
started repeat job https://jenkins.akka.io:8498/job/akka-multi-node-repeat/

File: akka-distributed-data/src/multi-jvm/scala/akka/cluster/ddata/ReplicatorMapDeltaSpec.scala [hunk 4]
```diff
@@ -97,23 +129,52 @@ object ReplicatorMapDeltaSpec extends MultiNodeConfig {
...
+  def removeElementFromORMap(om: ORMap[String, ORSet[String]], key: String, element: String)(implicit node: Cluster) =
+    om.updated(node, key, ORSet.empty[String])(_.remove(node, element))
+
+  def orsetOptionToSetOption(os: Option[ORSet[String]]): Option[Set[String]] =
+    os match {
+      case None ⇒ None
+      case Some(x) ⇒
+        val ORSet(s) = x
+        Some(s.asInstanceOf[Set[String]])
+    }
```
Comment (patriknw) [inline akka-distributed-data/src/multi-jvm/scala/akka/cluster/ddata/ReplicatorMapDeltaSpec.scala]: isn't all this `os.map(_.elements)`

**Review by patriknw — COMMENTED** [review]
looking good, I'll take another look tomorrow morning

File: akka-distributed-data/src/multi-jvm/scala/akka/cluster/ddata/ReplicatorMapDeltaSpec.scala (hunk 4 above)
Comment (gosubpl) [inline akka-distributed-data/src/multi-jvm/scala/akka/cluster/ddata/ReplicatorMapDeltaSpec.scala]: indeed, thanks!

File: akka-distributed-data/src/multi-jvm/scala/akka/cluster/ddata/ReplicatorMapDeltaSpec.scala [hunk 5]
```diff
@@ -230,18 +306,42 @@ class ReplicatorMapDeltaSpec extends MultiNodeSpec(ReplicatorMapDeltaSpec) with
...
+              }
+            case RemoveNoVD(key, elem, consistency) ⇒
+              runOn(first) {
+                fullStateReplicator ! Update(key._1, ORMultiMap.empty[String, String], WriteLocal)(_ removeBinding (key._2, elem))
+                deltaReplicator ! Update(key._1, ORMultiMap.empty[String, String], WriteLocal)(_ removeBinding (key._2, elem))
+              }
+            case AddOM(key, elem, consistency) ⇒
+              // to have an deterministic result when mixing add/remove we can only perform
+              // the ORSet operations from one node
+              runOn((if (key == KeyI) List(first) else List(first, second, third)): _*) {
```
Comment (patriknw) [inline akka-distributed-data/src/multi-jvm/scala/akka/cluster/ddata/ReplicatorMapDeltaSpec.scala]: This should be `key == KeyL`

Comment (gosubpl) [inline akka-distributed-data/src/multi-jvm/scala/akka/cluster/ddata/ReplicatorMapDeltaSpec.scala]: thanks for spotting, indeed.

**Comment by patriknw** [comment]
I ran ReplicatorMapDeltaSpec on repeat locally (with the KeyL change) and two failures, one for KeyD (see err1.txt) and one for KeyG (err2.txt). The latter is logging "Skipping DeltaPropagation ... from [G]"

[err1.txt](https://github.com/akka/akka/files/881495/err1.txt)
[err2.txt](https://github.com/akka/akka/files/881496/err2.txt)

**Comment by patriknw** [comment]
Here is one more that is rather easy to follow.
remove a
add f
add g
add d

It looks like the inital remove results in "a" being added in the end, for deltaReplicator

[err3.txt](https://github.com/akka/akka/files/881868/err3.txt)

**Comment by gosubpl** [comment]
err2.txt error is not reproducible. Will work now on err3.txt and err1.txt

**Comment by gosubpl** [comment]
also err3.txt input is working for me.

**Comment by gosubpl** [comment]
err3 sample input:
```
  val operations1 = Vector(AddVD(KeyE, "a", WriteLocal), Delay(386), AddOM(KeyJ, "b", WriteLocal), RemoveVD(KeyF, "a", WriteLocal), RemoveNoVD(KeyI, "b", WriteLocal), Delay(186), AddOM(KeyK, "e", WriteLocal), RemoveOM(KeyL, "e", WriteLocal), AddOM(KeyJ, "j", WriteLocal), Decr(KeyA, 4, WriteLocal), Decr(KeyC, 8, WriteMajority(5 seconds, 0)), AddVD(KeyF, "f", WriteLocal), Incr(KeyA, 78, WriteLocal), AddVD(KeyE, "c", WriteLocal), Decr(KeyC, 2, WriteLocal), Decr(KeyA, 6, WriteLocal), AddVD(KeyF, "g", WriteLocal), Decr(KeyC, 9, WriteLocal), Incr(KeyB, 13, WriteLocal), AddOM(KeyK, "g", WriteLocal), AddNoVD(KeyH, "a", WriteLocal), AddVD(KeyD, "d", WriteLocal), AddVD(KeyF, "d", WriteLocal), AddNoVD(KeyH, "h", WriteLocal), Decr(KeyA, 7, WriteLocal), Delay(158), Decr(KeyA, 5, WriteLocal), Incr(KeyB, 67, WriteLocal), AddOM(KeyK, "c", WriteLocal), Decr(KeyC, 4, WriteLocal), AddOM(KeyK, "j", WriteLocal), Incr(KeyC, 54, WriteLocal), Incr(KeyC, 63, WriteLocal), Delay(145), AddOM(KeyJ, "i", WriteLocal), Decr(KeyB, 4, WriteLocal), Incr(KeyC, 82, WriteLocal), Decr(KeyA, 9, WriteMajority(5 seconds, 0)), Incr(KeyB, 98, WriteLocal), Delay(284), Decr(KeyB, 8, WriteLocal), Incr(KeyC, 29, WriteLocal), AddNoVD(KeyG, "d", WriteLocal), AddVD(KeyD, "g", WriteLocal), AddOM(KeyJ, "c", WriteLocal), AddVD(KeyD, "j", WriteLocal), Delay(449), Delay(227), AddNoVD(KeyH, "j", WriteLocal), Delay(413), AddNoVD(KeyH, "c", WriteLocal), Incr(KeyC, 58, WriteLocal), Incr(KeyB, 80, WriteLocal), AddOM(KeyK, "i", WriteTo(2, 5 seconds)), Incr(KeyB, 87, WriteLocal), Delay(111), Decr(KeyA, 5, WriteLocal), AddNoVD(KeyG, "a", WriteLocal), Incr(KeyC, 5, WriteLocal), AddOM(KeyL, "c", WriteLocal), Decr(KeyA, 5, WriteLocal), Decr(KeyC, 4, WriteLocal), Delay(226), Delay(31), Delay(60), AddNoVD(KeyH, "b", WriteLocal), Incr(KeyC, 62, WriteLocal), AddOM(KeyL, "h", WriteLocal), Delay(7), AddNoVD(KeyG, "j", WriteLocal))

  val operations2 = Vector(Incr(KeyB, 12, WriteTo(2, 5 seconds)), Delay(441), AddOM(KeyJ, "j", WriteLocal), AddVD(KeyF, "e", WriteLocal), Delay(340), AddVD(KeyE, "f", WriteLocal), Incr(KeyC, 91, WriteLocal), Decr(KeyA, 9, WriteLocal), Incr(KeyC, 35, WriteTo(2, 5 seconds)), AddNoVD(KeyH, "e", WriteLocal), AddOM(KeyK, "d", WriteLocal), AddNoVD(KeyG, "d", WriteLocal), Delay(316), Decr(KeyA, 6, WriteLocal), Incr(KeyC, 13, WriteLocal), Incr(KeyA, 76, WriteLocal), Decr(KeyB, 9, WriteMajority(5 seconds, 0)), AddVD(KeyE, "b", WriteLocal), Decr(KeyA, 3, WriteLocal), Decr(KeyA, 2, WriteMajority(5 seconds, 0)), AddNoVD(KeyG, "d", WriteLocal), AddOM(KeyL, "c", WriteLocal), Delay(16), AddNoVD(KeyG, "e", WriteLocal), Incr(KeyB, 99, WriteLocal), Incr(KeyB, 49, WriteTo(2, 5 seconds)), AddVD(KeyF, "f", WriteLocal), Incr(KeyB, 9, WriteLocal), AddOM(KeyL, "g", WriteLocal), AddOM(KeyK, "j", WriteLocal), Delay(405), AddOM(KeyL, "b", WriteLocal), Incr(KeyC, 85, WriteTo(2, 5 seconds)), Decr(KeyA, 2, WriteLocal), AddNoVD(KeyG, "g", WriteLocal), Decr(KeyB, 0, WriteLocal), Delay(473), AddVD(KeyE, "j", WriteLocal), Incr(KeyB, 18, WriteTo(2, 5 seconds)), Decr(KeyA, 3, WriteLocal), Incr(KeyA, 8, WriteLocal), Delay(103), AddVD(KeyD, "b", WriteLocal), AddVD(KeyE, "e", WriteLocal), AddNoVD(KeyI, "j", WriteLocal), AddNoVD(KeyH, "c", WriteLocal), Incr(KeyA, 8, WriteTo(2, 5 seconds)), AddOM(KeyJ, "d", WriteLocal), AddVD(KeyE, "d", WriteLocal), Decr(KeyC, 7, WriteLocal), Decr(KeyC, 6, WriteLocal), Incr(KeyB, 59, WriteLocal), Incr(KeyA, 1, WriteLocal), AddVD(KeyD, "d", WriteLocal), Incr(KeyC, 25, WriteLocal), AddNoVD(KeyI, "j", WriteLocal), Decr(KeyB, 6, WriteTo(2, 5 seconds)), AddNoVD(KeyG, "a", WriteLocal), AddVD(KeyD, "c", WriteLocal), AddNoVD(KeyG, "i", WriteLocal), Delay(463))

  val operations3 = Vector(Delay(453), AddNoVD(KeyH, "a", WriteLocal), Incr(KeyC, 49, WriteLocal), Decr(KeyC, 4, WriteLocal), AddOM(KeyL, "d", WriteLocal), AddVD(KeyF, "d", WriteLocal), Decr(KeyA, 4, WriteTo(2, 5 seconds)), AddOM(KeyK, "a", WriteLocal), AddOM(KeyK, "e", WriteLocal), AddNoVD(KeyI, "d", WriteMajority(5 seconds, 0)), AddVD(KeyE, "f", WriteLocal), Incr(KeyC, 8, WriteLocal), AddVD(KeyD, "f", WriteLocal), Decr(KeyA, 7, WriteLocal), Decr(KeyA, 5, WriteLocal), Decr(KeyA, 3, WriteLocal), AddVD(KeyF, "d", WriteLocal), AddOM(KeyL, "f", WriteLocal), AddVD(KeyD, "g", WriteLocal), AddVD(KeyE, "e", WriteLocal), AddOM(KeyJ, "j", WriteLocal), Incr(KeyB, 29, WriteTo(2, 5 seconds)), AddVD(KeyD, "c", WriteLocal), Delay(401), AddNoVD(KeyI, "f", WriteLocal), Decr(KeyB, 7, WriteLocal), Incr(KeyB, 80, WriteLocal), Incr(KeyB, 56, WriteLocal), AddNoVD(KeyG, "f", WriteLocal), Delay(235), AddOM(KeyJ, "b", WriteLocal), AddVD(KeyD, "a", WriteLocal), AddOM(KeyJ, "c", WriteLocal), Decr(KeyC, 4, WriteLocal), Delay(46), AddNoVD(KeyH, "i", WriteLocal), Delay(227), AddOM(KeyL, "e", WriteLocal), Incr(KeyA, 93, WriteLocal), Decr(KeyA, 1, WriteLocal), AddOM(KeyK, "d", WriteLocal), AddNoVD(KeyH, "f", WriteLocal), AddOM(KeyJ, "b", WriteLocal), AddOM(KeyJ, "e", WriteLocal), AddVD(KeyE, "g", WriteTo(2, 5 seconds)), AddNoVD(KeyH, "f", WriteLocal), AddNoVD(KeyH, "j", WriteLocal), Incr(KeyA, 48, WriteTo(2, 5 seconds)), AddOM(KeyK, "j", WriteLocal), AddVD(KeyD, "i", WriteLocal), AddOM(KeyJ, "g", WriteTo(2, 5 seconds)), Decr(KeyB, 5, WriteLocal), AddNoVD(KeyG, "d", WriteLocal), AddVD(KeyE, "f", WriteLocal), Delay(89), Delay(300), Decr(KeyC, 8, WriteLocal), Delay(103), AddNoVD(KeyG, "c", WriteLocal), AddOM(KeyL, "a", WriteLocal), Incr(KeyB, 63, WriteTo(2, 5 seconds)), Decr(KeyA, 3, WriteLocal), Delay(334), Incr(KeyB, 29, WriteLocal), Delay(47), AddVD(KeyD, "c", WriteLocal), AddVD(KeyD, "j", WriteLocal), AddNoVD(KeyG, "e", WriteLocal))

  val operations4 = Vector(AddVD(KeyE, "e", WriteLocal), AddVD(KeyF, "a", WriteLocal), AddVD(KeyE, "a", WriteLocal), AddNoVD(KeyH, "j", WriteLocal), Incr(KeyA, 11, WriteLocal), AddOM(KeyL, "i", WriteLocal), Delay(360), AddOM(KeyL, "f", WriteLocal), Delay(72), AddOM(KeyJ, "d", WriteLocal), AddVD(KeyF, "b", WriteLocal), AddVD(KeyE, "i", WriteLocal), Decr(KeyC, 3, WriteMajority(5 seconds, 0)), Decr(KeyB, 8, WriteLocal), Delay(104), Delay(128), Decr(KeyB, 0, WriteLocal), Decr(KeyB, 5, WriteTo(2, 5 seconds)), AddVD(KeyF, "a", WriteLocal), AddVD(KeyF, "f", WriteLocal), Decr(KeyC, 4, WriteLocal), Decr(KeyC, 4, WriteLocal), AddOM(KeyL, "c", WriteLocal), Incr(KeyB, 72, WriteTo(2, 5 seconds)), AddOM(KeyK, "h", WriteTo(2, 5 seconds)), Delay(471), AddVD(KeyF, "d", WriteLocal), AddNoVD(KeyH, "g", WriteLocal), Delay(76), Decr(KeyA, 7, WriteLocal), Decr(KeyB, 6, WriteLocal), Delay(126), Decr(KeyB, 1, WriteLocal), AddNoVD(KeyI, "a", WriteLocal), Incr(KeyA, 11, WriteLocal), AddNoVD(KeyH, "j", WriteLocal), Decr(KeyB, 7, WriteLocal), Incr(KeyA, 29, WriteTo(2, 5 seconds)), Incr(KeyA, 99, WriteLocal), AddVD(KeyF, "d", WriteLocal), AddVD(KeyE, "c", WriteLocal), Decr(KeyA, 8, WriteLocal), Incr(KeyB, 74, WriteLocal), AddNoVD(KeyH, "j", WriteLocal), Decr(KeyC, 2, WriteLocal), Decr(KeyC, 2, WriteLocal), AddOM(KeyJ, "e", WriteLocal), AddVD(KeyF, "f", WriteLocal), Incr(KeyC, 35, WriteLocal), AddNoVD(KeyG, "j", WriteLocal), Decr(KeyA, 9, WriteLocal), AddOM(KeyL, "e", WriteLocal), AddNoVD(KeyI, "i", WriteLocal), AddNoVD(KeyI, "h", WriteLocal), AddOM(KeyK, "f", WriteLocal), AddNoVD(KeyI, "h", WriteLocal), AddVD(KeyE, "d", WriteTo(2, 5 seconds)), Decr(KeyA, 4, WriteMajority(5 seconds, 0)), Incr(KeyA, 46, WriteLocal), AddVD(KeyE, "d", WriteTo(2, 5 seconds)), AddNoVD(KeyH, "e", WriteLocal), AddVD(KeyF, "h", WriteLocal), Delay(179))

  val genOps = List(first → operations1, second → operations2, third → operations3, fourth → operations4).toMap
```

usage:
```  
"be eventually consistent" in {
      val operations = genOps(myself) //generateOperations(onNode = myself)
      log.debug(s"random operations on [${myself.name}]: ${operations.mkString(", ")}")

```

**Comment by gosubpl** [comment]
@patriknw I have rewritten the `mergeDelta` function in `ORMap`. Hopefully it is fully correct now. Ran `ReplicatorMapDeltaSpec` around 60 times, including the reproducers for the failing cases.
Can you please check how this looks on your side? Will probably leave this running on repeat tomorrow during the day.

**Comment by gosubpl** [comment]
100 executions of `ReplicatorMapDeltaSpec` without issues so far. 50 with increased number of generated random operations.
@patriknw maybe I should close this one, and open another with changes re-applied on top of current master?

**Comment by patriknw** [comment]
you can rebase your branch with `-i` and squash and such, then push forced to this branch and the PR will be "just as new"

**Comment by patriknw** [comment]
but opening a new PR is also fine, if you got lost in git

**Comment by gosubpl** [comment]
Closing, superseded by https://github.com/akka/akka/pull/22668

**Pull request closed without merging**